File: Newton_Verfahren.cc
#include <iostream>
#include <cmath>

//Typ-Definition: Zeiger auf f: double -> double
typedef double (*ftype)(double);

//Numerisches Differenzieren (funktional)
double ableitung(const ftype f, double h, double xk){
    return (f(xk+h) - f(xk)) / h;
}

//Newton-Verfahren (prozedural, mit call-by-reference)
void newton(const ftype myfunction, double& xk){
    while(true){    //Endlosschleife            
        double fk = myfunction(xk);                         //Zuweisung
        if (fabs(fk) < 1e-6) break;                         //Sprungbedingung
        xk = xk - fk / ableitung(myfunction, 1e-5, xk);     //Update via Zuweisung
    }
}

//Newton-Verfahren im Hauptprogramm
int main(){
    double xk = 1.0;
    newton(sin, xk);
    printf("Nullstelle x=%f, f(x)=%f\n", xk, sin(xk));
    return 0;
}

File: Newton_Verfahren.hs
--Numerisches Differenzieren (als Funktion)
ableitung f h = \x-> (f(x+h) - f(x)) / h

--Newton Update:
--x_{k+1} = x_k -Df(x_k)^{1} f(x_k)
update x f = x - f x / ableitung f 1e-5 x

--Newton Verfahren:
newton f x = if abs (f x) < 1e-6
    then x else newton f (update x f)
    
